### PR TITLE
Add isSecure for RaoResult

### PR DESCRIPTION
**Pull request by bqth29**
**Please check if the PR fulfills these requirements**

- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem?**
No

**What kind of change does this PR introduce?**
New feature

**What is the new behavior (if this is a feature change)?**
RaoResult objects now provides 3 new methods that allow client to assess the RAO result security status (i.e. positive margin on all Cnecs) , for multiple instants of optimization, and separately for each of the Cnec kind (Flow, Voltage and Angle).
These methods are :

- boolean isSecure() -> Assess security on last optimisation instant, taking into account all Cnec kinds
- boolean isSecure(PhysicalParameter...) -> Assess security on last optimisation instant, taking into account a potential subset of Cnec kinds
- boolean isSecure(Instant, PhysicalParameter...) -> Assess security on a given optimisation instant, taking into account a potential subset of Cnec kinds

This PR also add a method getLastInstant() to the CRAC API that allow to get the last instant based on order (biggest order one). This instant should correspond to the final optimisation instant.

**Does this PR introduce a breaking change or deprecate an API?**
- [ ] Yes
- [x] No

### CODE REVIEW

File: ra-optimisation/search-tree-rao/src/main/java/com/powsybl/openrao/searchtreerao/result/impl/PreventiveAndCurativesRaoResultImpl.java [hunk 1]
```diff
@@ -585,6 +586,19 @@ public void setOptimizationStepsExecuted(OptimizationStepsExecuted optimizationS
         }
     }
 
+    @Override
+    public boolean isSecure(Instant optimizedInstant, PhysicalParameter... u) {
+        if (ComputationStatus.FAILURE.equals(getComputationStatus())) {
+            return false;
+        }
+        return getCost(optimizedInstant) < 0;
```
Comment (pet-mit) [inline ra-optimisation/search-tree-rao/src/main/java/com/powsybl/openrao/searchtreerao/result/impl/PreventiveAndCurativesRaoResultImpl.java:594]: virtual costs may include MNEC costs, LoopFlow costs, etc that do not count in security status. If you want to check divergence, check the computation status map first like you did with computation status
```suggestion
        return getFunctionalCost(optimizedInstant) < 0;
```

File: data/rao-result/rao-result-impl/src/main/java/com/powsybl/openrao/data/raoresultimpl/RaoResultImpl.java [hunk 2]
```diff
@@ -339,6 +340,42 @@ public void setOptimizationStepsExecuted(OptimizationStepsExecuted optimizationS
...
+                    }
+                }
+            }
+        }
+        return true;
+    }
+
+    @Override
+    public boolean isSecure(Instant optimizedInstant, PhysicalParameter... u) {
+        if (ComputationStatus.FAILURE.equals(getComputationStatus())) {
```
Comment (pet-mit) [inline data/rao-result/rao-result-impl/src/main/java/com/powsybl/openrao/data/raoresultimpl/RaoResultImpl.java:368]: also check computation status map

**Review by pet-mit — CHANGES_REQUESTED** [review]
Please update the documentation here: https://github.com/farao-community/farao-community.github.io/blob/master/pages/docs/output-data/rao-result/rao-result-json.md

i realize this is more complicated than it seemed. we may want to do some refactoring in the future. it would be nice to have a review from @phiedw too

File: data/rao-result/rao-result-impl/src/main/java/com/powsybl/openrao/data/raoresultimpl/RaoResultImpl.java [hunk 3]
```diff
@@ -339,6 +340,42 @@ public void setOptimizationStepsExecuted(OptimizationStepsExecuted optimizationS
...
+    private boolean instantHasNoNegativeMargin(Instant optimizedInstant, PhysicalParameter... u) {
+        for (PhysicalParameter physicalParameter : Set.of(u)) {
+            switch (physicalParameter) {
+                case ANGLE -> {
+                    if (angleCnecResults.keySet().stream().anyMatch(cnec -> getMargin(optimizedInstant, cnec, Unit.DEGREE) < 0)) {
+                        return false;
+                    }
+                }
+                case FLOW -> {
+                    if (flowCnecResults.keySet().stream().anyMatch(cnec -> getMargin(optimizedInstant, cnec, Unit.MEGAWATT) < 0)) {
```
[the remark this answers is not in the document]
Comment (murgeyseb) [reply data/rao-result/rao-result-impl/src/main/java/com/powsybl/openrao/data/raoresultimpl/RaoResultImpl.java:352]: The problem here is that there is nothing enforced in the API that tell that functional cost only refer to flow cnecs (even if in practice it is currently the case)

**Review by pet-mit — APPROVED** [review]

**Pull request merged**